src/psyclone/tests/psyGen_test.py

```diff
@@ -1141,10 +1142,9 @@ def test_named_invoke_name_clash(tmpdir):
                            api="lfric")
     psy = PSyFactory("lfric", distributed_memory=True).create(invoke_info)
     gen = str(psy.gen)
-
-    assert ("SUBROUTINE invoke_a(invoke_a_1, b, istp, rdt, d, e, ascalar, "
+    assert ("SUBROUTINE invoke_a(invoke_a, b, istp, rdt, d, e, ascalar, "
             "f, c, g, qr)") in gen
-    assert "TYPE(field_type), intent(in) :: invoke_a_1" in gen
+    assert "TYPE(field_type), intent(in) :: invoke_a" in gen
```

Comment:
Ah damn, this has broken again since the `RoutineSymbol` isn't in the Routine again, as we remove the Routine's symbol from itself if it has a parent.
I'm not sure I have a good solution to prevent this behaviour - the previous solution of adding the RoutineSymbol into the routine broke for return symbols. 
Any ideas @sergisiso - we could add something into symbol table to prevent the same name for a symbol in a `Routine` but I'm not sure how we allow it for return symbols - it could also happen at lowering where we rename any symbol that shares a name with the Routine if its not the return symbol?

Reply:
This gets solved (this change can be reverted and no other tests fail) by removing the: `if self.parent is None and self.symbol_table:` line in `Routine.__init__` (the self.symbol_table is not needed because this is guaranteed by the super init). Something like this:

```
-        # If the routine is detached we try to
-        # add the symbol into it's symbol table unless the symbol table
-        # already contains a symbol with that name (e.g. for a Function).
-        if self.parent is None and self.symbol_table:
-            try:
-                self.symbol_table.lookup(self._symbol.name)
-            except KeyError:
-                self.symbol_table.add(self._symbol)
+        # Add the symbol into the routine itself, unless the symbol table
+        # already contains a symbol with that name (e.g. for a Function)
+        if self._symbol.name not in self.symbol_table:
+            self.symbol_table.add(self._symbol)
```

This means that we are always adding itself in the subroutine, in addition to the parent module when this is connected. I also checked that functions (with and without result clause) end up having the right symbols (wihout result the routinesymbols is replaced with a datasymbol). Do you think this is a feasible solution? Or are there drawbacks?

I am unsure if we could also simplify `update_parent_symbol_table` after this change.

Reply:
My fear with this was that it broke functions, but since they're ok I'm happy to make this change.